contracts/Safe.sol

```diff
@@ -411,23 +423,49 @@ contract Safe is
         address gasToken,
         address refundReceiver,
         uint256 _nonce
-    ) private view returns (bytes memory) {
-        bytes32 safeTxStructHash = keccak256(
-            abi.encode(
-                SAFE_TX_TYPEHASH,
-                to,
-                value,
-                keccak256(data),
-                operation,
-                safeTxGas,
-                baseGas,
-                gasPrice,
-                gasToken,
-                refundReceiver,
-                _nonce
-            )
-        );
-        return abi.encodePacked(bytes1(0x19), bytes1(0x01), domainSeparator(), safeTxStructHash);
+    ) private view returns (bytes memory result) {
+        bytes32 separatorHash = domainSeparator();
+
+        // The encoding is written in assembly to avoid memory re-allocations.
+        // Straightforward solidity implementation would allocate multiple times, but since the only reason it allocates is because it needs to hash the data,
+        // we can save gas by reusing the same memory block by overwriting the data.
+        /* solhint-disable no-inline-assembly */
+        assembly {
+            let freeMemoryPointer := mload(0x40)
+
+            // Copy the data from calldata to memory and calculate the hash
+            let calldataLength := data.length
+            calldatacopy(freeMemoryPointer, data.offset, calldataLength)
+            let calldataHash := keccak256(freeMemoryPointer, calldataLength)
+
+            // Prepare the SafeTX struct for hashing, overwriting the copied calldata
+            mstore(freeMemoryPointer, SAFE_TX_TYPEHASH)
+            mstore(add(freeMemoryPointer, 32), to)
+            mstore(add(freeMemoryPointer, 64), value)
+            mstore(add(freeMemoryPointer, 96), calldataHash)
+            mstore(add(freeMemoryPointer, 128), operation)
+            mstore(add(freeMemoryPointer, 160), safeTxGas)
+            mstore(add(freeMemoryPointer, 192), baseGas)
+            mstore(add(freeMemoryPointer, 224), gasPrice)
+            mstore(add(freeMemoryPointer, 256), gasToken)
+            mstore(add(freeMemoryPointer, 288), refundReceiver)
+            mstore(add(freeMemoryPointer, 320), _nonce)
+
+            // Hash the SafeTX struct and store it at the end of the result
+            // Hashing first so we can re-use the same memory block for the result
+            mstore(add(freeMemoryPointer, 66), keccak256(freeMemoryPointer, 352))
+            // Store the domain separator
+            mstore(add(freeMemoryPointer, 34), separatorHash)
+            // Store the EIP-712 prefix (0x1901)
+            mstore8(add(freeMemoryPointer, 33), 0x01)
+            mstore8(add(freeMemoryPointer, 32), 0x19)
+            // Store the length of the encoded data (always 66 bytes)
+            mstore(freeMemoryPointer, 66)
```

Comment:
I think you need to `mstore(0x40, add(freeMemoryPointer, 128))` (128 to allocate with padding, as Solidity does by default).

Otherwise, future allocations will clobber the data you prepared.